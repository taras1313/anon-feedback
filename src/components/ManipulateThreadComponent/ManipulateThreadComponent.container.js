import { connect } from 'react-redux';
import { bindActionCreators, compose } from 'redux';
import { withRouter } from 'react-router-dom';

import { ManipulateThreadComponent as ManipulateThread } from './ManipulateThreadComponent';
import {
  updateThreadField,
  resetThreadFields,
  createThread,
  editThread,
  setSelectedThread
} from '../../actions/threadActions';

const mapStateToProps = ({ threadsReducer: { threadData } }) => ({ threadData });

const mapDispatchToProps = (dispatch) => ({
  actions: bindActionCreators({
    updateThreadField,
    resetThreadFields,
    createThread,
    editThread,
    setSelectedThread
  }, dispatch)
});

export const ManipulateThreadComponent = compose(
  withRouter,
  connect(mapStateToProps, mapDispatchToProps)
)(ManipulateThread);

