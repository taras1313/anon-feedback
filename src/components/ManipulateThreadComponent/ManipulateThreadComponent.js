import React, { useState, useCallback } from 'react';
import { makeStyles } from '@material-ui/core/styles';
import Button from '@material-ui/core/Button';
import Dialog from '@material-ui/core/Dialog';
import TextField from '@material-ui/core/TextField';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import Checkbox from '@material-ui/core/Checkbox';
import AppBar from '@material-ui/core/AppBar';
import Toolbar from '@material-ui/core/Toolbar';
import IconButton from '@material-ui/core/IconButton';
import Typography from '@material-ui/core/Typography';
import CloseIcon from '@material-ui/icons/Close';
import Slide from '@material-ui/core/Slide';
import cx from 'classnames';
import { uniqueNamesGenerator, adjectives, animals } from 'unique-names-generator';

const EDIT = 'edit';
const CREATE = 'create';

const useStyles = makeStyles(theme => ({
  root: {
    width: '80%',
    height: '70vh'
  },
  appBar: {
    position: 'relative',
    backgroundColor: '#424b5f',
    boxShadow: 'none'
  },
  title: {
    marginLeft: theme.spacing(2),
    textTransform: 'capitalize',
    flex: 1,
  },
  container: {
    display: 'flex',
    flexWrap: 'wrap',
    padding: '30px 40px',
    flexDirection: 'column',
    flex: 1
  },
  textField: {
    marginLeft: theme.spacing(1),
    marginRight: theme.spacing(1),
    width: '100%',
    whiteSpace: 'pre-line'
  },
  smallTextField: {
    width: 200
  },
  nickNameWrapper: {
    display: 'flex',
    flexDirection: 'column'
  },
  formControl: {
    margin: 0
  },
  controlBtns: {
    marginTop: '16px',
    marginLeft: 'auto',
    textTransform: 'capitalize'
  },
  button: {
    fontWeight: 'bold',
    boxShadow: 'none',
    marginLeft: '16px',
    fontSize: '0.7rem'
  },
  submitBtn: {
    backgroundColor: '#65c178',
    fontWeight: 'bold',
    color: 'white',
    '&:hover': {
      backgroundColor: '#159776'
    },
  }
}));

const Transition = React.forwardRef(function Transition(props, ref) {
  return <Slide direction="up" ref={ref} {...props} />;
});

export const ManipulateThreadComponent = (
  {
    history,
    action,
    isOpen,
    onClose,
    actions: {
      updateThreadField,
      resetThreadFields,
      createThread,
      editThread,
      setSelectedThread
    },
    threadData: {
      title,
      text,
      username
    }
  }) => {
  const classes = useStyles();
  const [isChecked, setIsChecked] = useState(false);

  const toggleAutoGenerator = () => {
    const config = {
      dictionaries: [adjectives, animals],
      separator: ' ',
      length: 2,
    };

    const value = !isChecked ? uniqueNamesGenerator(config) : '';

    updateThreadField({ field: 'username', value });
    setIsChecked(!isChecked);
  };

  const updateThreadData = ({ target: { value, name } }) => {console.log(value.toString());updateThreadField({ field: name, value })};

  const manipulateThread = () => {
    switch (action) {
      case EDIT: {
        editThread().then(data => {
          setSelectedThread(data);
          onClose();
        });
        break;
      }

      case CREATE: {
        createThread().then(data => {
          const { _id } = data;
          const { push } = history;

          setSelectedThread(data);
          push(`/feed/${_id}`);
          onClose();
        });
        break;
      }

      default:
        return;
    }
  };

  return (
    <div>
      <Dialog
        fullWidth
        maxWidth="md"
        open={isOpen}
        onClose={onClose}
        TransitionComponent={Transition}
      >
        <AppBar className={classes.appBar}>
          <Toolbar>
            <IconButton edge="start" color="inherit" onClick={onClose} aria-label="close">
              <CloseIcon />
            </IconButton>
            <Typography variant="h6" className={classes.title}>
              {action} Thread
            </Typography>
          </Toolbar>
        </AppBar>
        <div className={classes.container}>
          {action === CREATE && (
            <div className={classes.nickNameWrapper}>
              <TextField
                autoComplete='off'
                id="standard-basic"
                className={cx(classes.textField, classes.smallTextField)}
                onChange={updateThreadData}
                required
                value={username}
                name="username"
                label="Anon nick name"
              />
              <FormControlLabel
                className={classes.formControl}
                control={
                  <Checkbox
                    checked={isChecked}
                    onChange={toggleAutoGenerator}
                    color="primary"
                  />
                }
                label="Generate nick name for you?"
              />
            </div>
          )}

          <TextField
            autoComplete='off'
            required
            id="standard-multiline-flexible"
            label="Thread Title"
            rows="2"
            multiline
            value={title}
            onChange={updateThreadData}
            name="title"
            className={classes.textField}
            margin="normal"
            variant="outlined" // ?
            // variant="filled" // ?
          />

          <TextField
            required
            autoComplete='off'
            id="standard-multiline-flexible"
            label="Thread Description"
            multiline
            rows="6"
            value={text}
            onChange={updateThreadData}
            name="text"
            className={classes.textField}
            margin="normal"
            variant="outlined" // ?
            // variant="filled" // ?

          />

          <div className={classes.controlBtns}>
            <Button variant="contained" onClick={resetThreadFields} color="secondary" className={classes.button}>
              Reset
            </Button>

            <Button variant="contained" onClick={manipulateThread} className={cx(classes.button, classes.submitBtn)}>
              {action} Thread
            </Button>
          </div>
        </div>
      </Dialog>
    </div>
  );
};