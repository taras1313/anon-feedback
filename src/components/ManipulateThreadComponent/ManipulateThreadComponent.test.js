import React from "react";
import { ManipulateThreadComponent } from "./ManipulateThreadComponent";
import FormControlLabel from "@material-ui/core/FormControlLabel";
import TextField from "@material-ui/core/TextField";
import Button from "@material-ui/core/Button";
import { shallow } from "enzyme";
// import { createMemoryHistory } from 'history';

// const history = createMemoryHistory();

describe("threadAuditComponent", () => {
  const props = {
    onClose: () => {},
    actions: {
      updateThreadField: jest.fn(),
      createThread: jest
        .fn()
        .mockImplementation(() =>
          Promise.resolve({ thread: "testThread", user: "testUser" })
        ),
      editThread: jest
        .fn()
        .mockImplementation(() => Promise.resolve("testData")),
      setSelectedThread: jest.fn(),
      setUser: jest.fn(),
    },
    history: {
      push: jest.fn(),
    },
    action: "create",
  };

  const useState = jest.spyOn(React, "useState");

  afterEach(() => {
    jest.clearAllMocks();
  });

  test("renders w/o props", () => {
    const wrapper = shallow(<ManipulateThreadComponent />);

    expect(wrapper).toMatchSnapshot();
  });

  test("autocomplete", () => {
    const setIsChecked = jest.fn();
    useState.mockImplementationOnce(() => [true, setIsChecked]);
    const wrapper = shallow(<ManipulateThreadComponent {...props} />);
    wrapper.find(FormControlLabel).props().control.props.onChange();
    expect(props.actions.updateThreadField).toHaveBeenCalledTimes(1);
    expect(setIsChecked).toHaveBeenCalledWith(false);
  });

  test("autocomplete", () => {
    const setIsChecked = jest.fn();
    useState.mockImplementationOnce(() => [false, setIsChecked]);
    const wrapper = shallow(<ManipulateThreadComponent {...props} />);
    wrapper.find(FormControlLabel).props().control.props.onChange();
    expect(props.actions.updateThreadField).toHaveBeenCalledTimes(1);
    expect(setIsChecked).toHaveBeenCalledWith(true);
  });

  test("text field", () => {
    const wrapper = shallow(<ManipulateThreadComponent {...props} />);
    wrapper
      .find(TextField)
      .at(1)
      .props()
      .onChange({
        target: {
          value: "test",
          name: "test",
        },
      });
    expect(props.actions.updateThreadField).toHaveBeenCalledTimes(1);
  });

  test("manipulate thread create", async () => {
    const wrapper = shallow(<ManipulateThreadComponent {...props} />);

    wrapper.find(Button).at(1).props().onClick();

    await setTimeout(() => {}, 1);
    expect(props.actions.setUser).toHaveBeenCalledTimes(1);
  });

  test("manipulate thread edit", async () => {
    const wrapper = shallow(
      <ManipulateThreadComponent {...props} action="edit" />
    );

    wrapper.find(Button).at(1).props().onClick();

    await setTimeout(() => {}, 1);
    expect(props.actions.setSelectedThread).toHaveBeenCalledWith("testData");
  });

  test("manipulate thread edit", async () => {
    const wrapper = shallow(
      <ManipulateThreadComponent {...props} action="vbrost" />
    );

    wrapper.find(Button).at(1).props().onClick();
  });

  test("default methods", () => {
    const wrapper = shallow(<ManipulateThreadComponent action='create'/>);

    wrapper.find(Button).at(0).props().onClick();
    wrapper.find(FormControlLabel).props().control.props.onChange();
  });
});
