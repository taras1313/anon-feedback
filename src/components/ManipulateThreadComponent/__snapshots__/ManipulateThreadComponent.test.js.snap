// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`threadAuditComponent renders w/o props 1`] = `
<div>
  <WithStyles(ForwardRef(Dialog))
    TransitionComponent={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "render": [Function],
      }
    }
    fullWidth={true}
    maxWidth="md"
  >
    <WithStyles(ForwardRef(AppBar))
      className="makeStyles-appBar-2"
    >
      <WithStyles(ForwardRef(Toolbar))>
        <WithStyles(ForwardRef(IconButton))
          aria-label="close"
          color="inherit"
          edge="start"
        >
          <CloseIcon />
        </WithStyles(ForwardRef(IconButton))>
        <WithStyles(ForwardRef(Typography))
          className="makeStyles-title-3"
          variant="h6"
        >
           Thread
        </WithStyles(ForwardRef(Typography))>
      </WithStyles(ForwardRef(Toolbar))>
    </WithStyles(ForwardRef(AppBar))>
    <div
      className="makeStyles-container-4"
    >
      <WithStyles(ForwardRef(TextField))
        autoComplete="off"
        className="makeStyles-textField-5"
        id="standard-multiline-flexible"
        label="Thread Title"
        margin="normal"
        multiline={true}
        name="title"
        onChange={[Function]}
        required={true}
        rows="2"
        variant="outlined"
      />
      <WithStyles(ForwardRef(TextField))
        autoComplete="off"
        className="makeStyles-textField-5"
        id="standard-multiline-flexible"
        label="Thread Description"
        margin="normal"
        multiline={true}
        name="text"
        onChange={[Function]}
        required={true}
        rows="6"
        variant="outlined"
      />
      <div
        className="makeStyles-controlBtns-9"
      >
        <WithStyles(ForwardRef(Button))
          className="makeStyles-button-10"
          color="secondary"
          onClick={[Function]}
          variant="contained"
        >
          Reset
        </WithStyles(ForwardRef(Button))>
        <WithStyles(ForwardRef(Button))
          className="makeStyles-button-10 makeStyles-submitBtn-11"
          onClick={[Function]}
          variant="contained"
        >
           Thread
        </WithStyles(ForwardRef(Button))>
      </div>
    </div>
  </WithStyles(ForwardRef(Dialog))>
</div>
`;
